src/systems/filecoin_blockchain/chainsync/_index.md

```diff
@@ -1,58 +1,388 @@
+        - (This is an extension and not required. can just do Step 4.)
+  - **Step 4. (receiver) `Pull Messages`**:
+      - if `receiver` **DOES NOT** already have the full block for `b.BlockCID`, then:
+        - if `receiver` has _some_ of the messages:
+          - `receiver` requests missing `Messages` and `MessageReceipts` from `sender`:
+              - `Graphsync.Pull(sender, Select(m3, m10, m50, ...))`
+        - if `receiver` does not have any of the messages (default safe but expensive thing to do):
+          - `receiver` requests all `Messages` and `MessageReceipts` from `sender`:
+              - `Graphsync.Pull(sender, SelectAll(bh.Messages), SelectAll(bh.MessageReceipts))`
+        - (This is the largest amount of stuff)
+  - **Step 5. (receiver) `Validate Block`**:
+      - the only remaining thing to do is to complete Block Validation.
+
+
+## ChainSync FSM: `INIT`
+
+- beginning state. no network connections, not synchronizing.
+- local state is loaded: internal data structures (eg chain, cache) are loaded
+- `LastTrustedCheckpoint` is set the latest network-wide accepted `TrustedCheckpoint`
+- **Chain State and Finality**:
+  - In this state, the **chain MUST NOT advance** beyond whatever the node already has.
+  - No new blocks are reported to consumers.
+  - The chain state provided is whatever was loaded from prior executions (worst case is `LastTrustedCheckpoint`)
+- **security conditions to transition out:**
+  - local state and data structures SHOULD be verified to be correct
+    - this means validating any parts of the chain or `StateTree` the node has, from `LastTrustedCheckpoint` on.
+  - `LastTrustedCheckpoint` is well-known across the Filecoin Network to be a true `TrustedCheckpoint`
+    - this SHOULD NOT be verified in software, it SHOULD be verified by operators
+    - Note: we ALWAYS have at least one `TrustedCheckpoint`, the `GenesisCheckpoint`.
+- **transitions out:**
+  - once done verifying things: move to `BOOTSTRAP`
+
+## ChainSync FSM: `BOOTSTRAP`
+
+- `network.Bootstrap()`: establish connections to peers until we satisfy security requirement
+  - for better security, use many different `libp2p.PeerDiscovery` protocols
+- `BlockPubsub.Bootstrap()`: establish connections to `BlockPubsub` peers
+- `Graphsync.Serve()`: set up a Graphsync service, that responds to others' queries
+- **Chain State and Finality**:
+  - In this state, the **chain MUST NOT advance** beyond whatever the node already has.
+  - No new blocks are reported to consumers.
+  - The chain state provided is whatever was loaded from prior executions (worst case is `LastTrustedCheckpoint`).
+- **security conditions to transition out:**
+  - `Network` connectivity MUST have reached the security level acceptable for `ChainSync`
+  - `BlockPubsub` connectivity MUST have reached the security level acceptable for `ChainSync`
```

Comment:
Are these security levels specified by the protocol? Seems like it could have network wide implications if everyone chooses lax security here but maybe not so its ok to leave up to user?